### PR TITLE
fix(core): consistent pnpm lockfile v6 version parsing

### LINKED ISSUE #16891: Broken PNPM Lock Files
### Current Behavior

When building a project that generates a lock file with pnpm it comes out broken (when trying to install it errors like the following are reported)

```
WARN  Broken lockfile: no entry for '/debug/4.3.4(supports-color@8.1.1)' in pnpm-lock.yaml
ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  The lockfile is broken! Resolution step will be performed to fix it.
```

### Expected Behavior

Lockfiles should contain all references such that they are considered to be valid by pnpm

### GitHub Repo

_No response_

### Steps to Reproduce

1. Build a project with a generated pnpm lockfile
2. Copy the dist built to another directory (or docker image)
3. `pnpm install`
4. Observe the behavior

Note: Issue depends on dependency tree of that project.

### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node   : 18.13.0
   OS     : darwin arm64
   pnpm   : 8.4.0
   Hasher : Native
   
   nx (global)        : 16.1.3
   nx                 : 16.1.3
   @nx/js             : 16.1.3
   @nx/jest           : 16.1.3
   @nx/linter         : 16.1.3
   @nx/workspace      : 16.1.3
   @nx/angular        : 16.1.3
   @nx/cypress        : 16.1.3
   @nx/detox          : 16.1.3
   @nx/devkit         : 16.1.3
   @nx/eslint-plugin  : 16.1.3
   @nx/nest           : 16.1.3
   @nx/next           : 16.1.3
   @nx/node           : 16.1.3
   @nx/plugin         : 16.1.3
   @nx/react          : 16.1.3
   @nx/react-native   : 16.1.3
   @nx/web            : 16.1.3
   @nx/webpack        : 16.1.3
   typescript         : 5.0.4
   ---------------------------------------
   Community plugins:
   @jscutlery/semver        : 3.0.0
   @ngneat/spectator        : 14.0.0
   @ngrx/effects            : 16.0.0
   @ngrx/entity             : 16.0.0
   @ngrx/router-store       : 16.0.0
   @ngrx/schematics         : 16.0.0
   @ngrx/store              : 16.0.0
   @ngrx/store-devtools     : 16.0.0
   @nx-tools/nx-container   : 4.0.3
   @rx-angular/cdk          : 15.0.0
   @rx-angular/template     : 15.2.0
   eslint-config-rocketyard : 0.1.0
   eslint-plugin-ngrx       : 2.1.4
   ng-mocks                 : 14.10.0
   ngx-deploy-npm           : 5.2.0
   nx-stylelint             : 15.0.0
   ---------------------------------------
   Local workspace plugins:
         @rocketyard/file-storage-nx
         @rocketyard/redis-nx
         @rocketyard/nats-nx
         @rocketyard/github-nx
         @rocketyard/pg-nx
         @flagship-create/flagship-nx
         @rocketyard/orchestrate-nx
         eslint-config-rocketyard
         @rocketyard/terraform-nx
```

### Failure Logs

_No response_

### Operating System

- [ ] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

This issue did not occur in 15

### PR DESCRIPTION
**Pull request by halfbakedsneed**
## Current Behavior

pnpm v8 bumped the lockfile format from 5 -> 6. This changed the way peer dependencies are encoded within lockfile version paths. https://github.com/nrwl/nx/pull/15925 added support for v6 lockfiles, but introduced some inconsistencies in the way package versions are parsed:

- During the the addition of external nodes to the graph, package versions are parsed correctly (switching on separator - `(` or `_` depending on lockfile version).
- During the derivation of hoisted versions, package versions are always treated as though the lockfile is <= v5 (using only the `_` separator to parse the version).
- During the addition of static dependencies to the graph, package versions are always treated as though the lockfile is <= v5 (using only the `_` separator to parse the version).

The result of these inconsistencies is that the the resulting project graph may omit necessary nodes.

## Expected Behavior

Package versions should be parsed consistently throughout the graph augmentation process.

## Related Issue(s)

Fixes https://github.com/nrwl/nx/issues/16891

### CODE REVIEW

File: packages/nx/src/plugins/js/lock-file/pnpm-parser.spec.ts [hunk 1]
```diff
@@ -183,12 +183,12 @@ describe('pnpm LockFile utility', () => {
         // this is our pruned lock file structure
         const prunedGraph = pruneProjectGraph(graph, appPackageJson);
         expect(Object.keys(prunedGraph.externalNodes).length).toEqual(
-          864 + 117 // peer cypress adds additional 117 deps
+          864 + 119 // peer cypress adds additional 119 deps
         );
 
         // this should not fail
         expect(() =>
-          stringifyPnpmLockfile(prunedGraph, lockFile, appPackageJson)
```
Comment (halfbakedsneed) [inline packages/nx/src/plugins/js/lock-file/pnpm-parser.spec.ts:191]: Currently on `main`, if actually try to `pnpm install --frozen-lockfile` using the stringified lockfile, you get a ` ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY `.

**Review by meeroslav — APPROVED** [review]
Looks good to me. Thank you!

**Comment by meeroslav** [comment]
The failing storybook test is not related to this PR. I will try to fix it in a separate PR and rebase this one.

**Comment by wSedlacek** [comment]
I am excited to test this.

**Pull request merged**